File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/coding_interviews_2/Question_0036_ConvertBinarySearchTree.java
package com.lanceolata.coding.algorithms.java.coding_interviews_2;

public class Question_0036_ConvertBinarySearchTree {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    private TreeNode last;

    public TreeNode Convert(TreeNode pRootOfTree) {
        if (pRootOfTree == null) {
            return null;
        }
        last = null;
        help(pRootOfTree);
        while (pRootOfTree.left != null) {
            pRootOfTree = pRootOfTree.left;
        }
        return pRootOfTree;
    }

    private void help(TreeNode root) {
        if (root == null) {
            return;
        }
        help(root.left);
        root.left = last;
        if (last != null) {
            last.right = root;
        }
        last = root;
        if (root.right != null) {
            help(root.right);
        }
    }
}

File: algorithms/algorithms-cplusplus/coding_interviews_2/Question_0036_ConvertBinarySearchTree.cc
#include <stack>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
 public:
  TreeNode* Convert(TreeNode* pRootOfTree) {
    if (!pRootOfTree) {
      return pRootOfTree;
    }
    TreeNode *pre = NULL;
    stack<TreeNode*> st;
    while (pRootOfTree || !st.empty()) {
      if (pRootOfTree) {
        st.push(pRootOfTree);
        pRootOfTree = pRootOfTree->left;
      } else {
        pRootOfTree = st.top();
        st.pop();
        if (pre) {
          pre->right = pRootOfTree;
          pRootOfTree->left = pre;
        }
        pre = pRootOfTree;
        pRootOfTree = pRootOfTree->right;
      }
    }
    while (pre->left) {
      pre = pre->left;
    }
    return pre;
  }

  TreeNode* ConvertRecursive(TreeNode* pRootOfTree) {
    if (!pRootOfTree) {
      return pRootOfTree;
    }
    TreeNode *pre = NULL;
    help(pRootOfTree, &pre);
    while (pRootOfTree->left) {
      pRootOfTree = pRootOfTree->left;
    }
    return pRootOfTree;
  }

  void help(TreeNode* root, TreeNode** pre) {
    if (!root) {
      return;
    }
    help(root->left, pre);
    if (*pre) {
      (*pre)->right = root;
      root->left = (*pre);
    }
    *pre = root;
    help(root->right, pre);
  }
};
